### PR TITLE
Token parameter in gateway client

### LINKED ISSUE #253: Add the token parameter to the gateway API methods

### PR DESCRIPTION
**Pull request by war-in**
### **Please check if the PR fulfills these requirements**
- [x] The formatter, linter, and tests all run without an error
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

## **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

(closes #253 )

* **What is the current behavior?** (You can also link to an open issue here)

### CODE REVIEW

File: starknet_py/net/gateway_client.py [hunk 1]
```diff
@@ -285,12 +295,21 @@ async def get_class_by_hash(self, class_hash: Hash) -> DeclaredContract:
...
     async def _add_transaction(
-        self, tx: StarknetTransaction
+        self,
+        tx: StarknetTransaction,
+        token: Optional[str] = None,
     ) -> SentTransactionResponse:
-        res = await self._gateway_client.post(
-            method_name="add_transaction",
-            payload=StarknetTransaction.Schema().dump(obj=tx),
-        )
```
Comment (cptartur) [inline starknet_py/net/gateway_client.py:293]: How about `params = {"token": token} if token is not None else {}` then you don't have to duplicate client.post(...) code.

**Review by cptartur — CHANGES_REQUESTED** [review]

**Review by cptartur — APPROVED** [review]
Looks good to me. However I'd like @Arcticae to also review this one before we merge.

**Comment by Arcticae** [comment]
This seems fine codewise but i'd rather have confirmation from SW that we really need this, and this is going to get used some day

File: starknet_py/net/gateway_client.py [hunk 2]
```diff
@@ -259,15 +259,25 @@ async def call_contract(
         return [int(v, 16) for v in res["result"]]
 
     async def send_transaction(
-        self, transaction: InvokeFunction
+        self,
+        transaction: InvokeFunction,
+        token: Optional[str] = None,
     ) -> SentTransactionResponse:
-        return await self._add_transaction(transaction)
+        return await self._add_transaction(transaction, token)
```
Comment (ca11ab1e) [inline starknet_py/net/gateway_client.py:266]:
```suggestion
        return await self._add_transaction(transaction, token=token)
```

File: starknet_py/net/gateway_client.py [hunk 3]
```diff
@@ -259,15 +259,25 @@ async def call_contract(
...
 
-    async def deploy(self, transaction: Deploy) -> SentTransactionResponse:
-        return await self._add_transaction(transaction)
+    async def deploy(
+        self,
+        transaction: Deploy,
+        token: Optional[str] = None,
+    ) -> SentTransactionResponse:
+        return await self._add_transaction(transaction, token)
```
Comment (ca11ab1e) [inline starknet_py/net/gateway_client.py:273]:
```suggestion
        return await self._add_transaction(transaction, token=token)
```

**Comment by cptartur** [comment]
> This seems fine codewise but i'd rather have confirmation from SW that we really need this, and this is going to get used some day

I think we should focus on releasing 0.9.1 and deal with this PR once SW replies then.

File: starknet_py/net/gateway_client.py [hunk 4]
```diff
@@ -259,15 +259,25 @@ async def call_contract(
...
 
-    async def declare(self, transaction: Declare) -> SentTransactionResponse:
-        return await self._add_transaction(transaction)
+    async def declare(
+        self,
+        transaction: Declare,
+        token: Optional[str] = None,
+    ) -> SentTransactionResponse:
+        return await self._add_transaction(transaction, token)
```
Comment (ca11ab1e) [inline starknet_py/net/gateway_client.py:280]:
```suggestion
        return await self._add_transaction(transaction, token=token)
```

**Comment by Arcticae** [comment]
> > This seems fine codewise but i'd rather have confirmation from SW that we really need this, and this is going to get used some day
> 
> I think then we should focus on releasing 0.9.1 and deal with this PR once SW replies.

That is fine by me

**Comment by ca11ab1e** [comment]
> > This seems fine codewise but i'd rather have confirmation from SW that we really need this, and this is going to get used some day
> 
> I think then we should focus on releasing 0.9.1 and deal with this PR once SW replies.

I can reply here: we will use it :)

**Comment by ca11ab1e** [comment]
And thanks for being fast to respond!

**Comment by ca11ab1e** [comment]
> > This seems fine codewise but i'd rather have confirmation from SW that we really need this, and this is going to get used some day
> 
> I think then we should focus on releasing 0.9.1 and deal with this PR once SW replies.

@cptartur please include tha tsmall patch in the same release as you will ship Cairo 0.9.1 :pray: 
Not only it would block SW to move forward, but also the whole Ape StarkNet tooling.

**Pull request merged**